### ISSUE TITLE
How can i develop a custom transfer protocol for meterpreter 

### ISSUE THREAD

**Issue by Feb30th** [rapid7/metasploit-framework#13580]
## Steps to reproduce

How can i develop a custom transfer protocol for meterpreter?

## scenario
I want to write up a handler named "windows/meterpreter_bind_http".
As the name suggests, I hope my meterperter binds on HTTP URI on target host and wait for my handler to establish connection to it

**Comment by FunnyWolf** [comment]
`metasploit-framework\lib\msf\core\handler\bind_tcp.rb`
`metasploit-payloads\c\meterpreter\source\metsrv\server_transport_tcp.c`
`metasploit-payloads\c\meterpreter\source\metsrv\server_transport_winhttp.c`
i think those files will help,but why not bind_tcp?for port reuse?

**Comment by OJ** [comment]
Adding transports is a non-trivial activity at the best of times. It requires quite a bit of effort on both sides of the wire (ie. MSF and Meterpreter). The example youv'e chosen effectively reverses the roles that MSF and Meterpreter have, meaning that you'd have to have MSF constantly poll Meterpreter and not the other way around. While the MSF side of this equation isn't that hard, the Meterpreter side will be hard.

Issues you'll face on the Meterpreter side are at least:
* Bind ports - on devices that have firewalls you'll have to somehow bypass them.
* If you want to reuse 80/443 you'll have to find some way of running elevated and/or wiring into existing web infrastructure.
* Handling command queuing and ordering so that the packets are reachable by MSF.
* Making sure packets are handled/processed in order on the Meterpreter side.

If you want to do this on existing infra that uses a webserver, you'll need implementations that can work with IIS,Coyote,Nginx,etc. If you target a Windows desktop you'll have to tap into IIS if it's there, or make use of direct `httpd.sys` calls via the API. Either way, expect pain, especially in the latter case where there's no firewall rule configured to allow traffic on the port.

In short, it's a huge job. Feel free to give it a go, but expect it to be a fight!
